doc/designs/quic-design/quic-ackm.md

```diff
@@ -0,0 +1,412 @@
+
+int QUIC_ACKM_on_rx_packet(QUIC_ACKM *ackm, const QUIC_ACKM_RX_PKT *pkt);
+```
+
+### On RX ACK Frame
+
+This must be called whenever an ACK frame is received. It should be called
+after any call to `QUIC_ACKM_on_rx_packet`.
+
+The ranges of packet numbers being acknowledged are passed as an argument.
+`pkt_space` is one of the `QUIC_PN_SPACE_*` values, specifying the packet number
+space of the containing packet. `rx_time` is the time the frame was
+received.
+
+This function causes `on_acked` callbacks to be invoked on applicable packets.
+
+Returns 1 on success.
+
+```c
+typedef struct quic_ackm_ack_range_st {
+    /*
+     * Represents an inclusive range of packet numbers [start, end].
+     * start must be <= end.
+     */
+    QUIC_PN start, end;
+} QUIC_ACKM_ACK_RANGE;
+
+typedef struct quic_ackm_ack {
+    /*
+     * A sequence of packet number ranges [[start, end]...].
+     *
+     * The ranges must be sorted in descending order, for example:
+     *      [ 95, 100]
+     *      [ 90,  92]
+     *      etc.
+     *
+     * As such, ack_ranges[0].end is always the highest packet number
+     * being acknowledged and ack_ranges[num_ack_ranges-1].start is
+     * aalways the lowest packet number being acknowledged.
+     *
+     * num_ack_ranges must be greater than zero, as an ACK frame must
+     * acknowledge at least one packet number.
+     */
+    const QUIC_ACKM_ACK_RANGE  *ack_ranges;
+    size_t                      num_ack_ranges;
+
+    OSSL_TIME                   delay_time;
+    uint64_t                    ect0, ect1, ecnce;
+} QUIC_ACKM_ACK;
+
+int QUIC_ACKM_on_rx_ack_frame(QUIC_ACKM *ackm, const QUIC_ACKM_ACK *ack,
+                              int pkt_space, OSSL_TIME rx_time);
+```
+
+### On Packet Space Discarded
+
+This must be called whenever a packet number space is discarded. ACK-tracking
+information for the number space is thrown away. Any previously provided
+`QUIC_ACKM_TX_PKT` structures have their `on_discarded` callback invoked,
+providing an opportunity for them to be freed.
+
+Returns 1 on success.
+
+```c
+int QUIC_ACKM_on_pkt_space_discarded(QUIC_ACKM *ackm, int pkt_space);
+```
+
+### On Handshake Connfirmed
+
+This should be called by the user when the QUIC handshake is confirmed. The PTO
+algorithm behaves differently depending on whether the QUIC handshake is
+confirmed yet.
+
+Returns 1 on success.
+
+```c
+int QUIC_ACKM_on_handshake_confirmed(QUIC_ACKM *ackm);
+```
+
+### On Timeout
+
+This must be called whenever the loss detection deadline expires.
+
+```c
+int QUIC_ACKM_on_timeout(QUIC_ACKM *ackm);
+```
+
+Queries
+-------
+
+These functions allow information about the status of the ACK manager to be
+obtained.
+
+### Get Loss Detection Timeout
+
+This returns a deadline after which `QUIC_ACKM_on_timeout` should be called.
+
+If it is `OSSL_TIME_NEVER`, no timeout is currently active.
+
+The value returned by this function may change after any call to any of the
+event functions above is made.
```

Comment:
I suppose the timeout will be a timed event in the event queue, does it mean it has to be constantly updated (i.e. moved around in the queue)? I suppose so. That has some implications for the event queue design @paulidale. Now the question is which component should do the updates of the event - to me it would be more natural to just directly put (and update) the event in the queue from the ack manager, although that would make it less isolated from the rest of the code.

One middle-ground option would be to add a callback that would be set when creating the ACK manager that would be called whenever the timeout event needs to be updated.

Reply:
The event queue has a postpone event call which could be used.  I think it would be the ACK Manager's task to update the event queue either using this or remove/add.

However, I see no reason why all timeout events couldn't be added to the event queue and this function ceasing to exist.  In fact, I'd favour this approach over an ad-hoc updating.  In other words, the ACK Manager adds and removes events for all packets from the event queue directly.